incipyt/project/environment.py

```diff
@@ -41,25 +75,45 @@ def __getitem__(self, key):
 
     def __setitem__(self, key, value):
         if key in self.data:
-            raise ValueError(f"Environment variable {key} already exists, delete it before.")
-
-        logger.debug("Set environment variable %s=%s.", key, value)
+            raise ValueError(f"Environment variable {key} is set already, delete it before.")
+        logger.debug("set %s=%s.", key, value)
         self.data[key] = value
 
     def _prompt(self, key):
-        if key in variables.metadata and variables.metadata[key].do_not_prompt:
-            user_input = variables.metadata[key].default
-        else:
-            user_input = click.prompt(
-                key.replace("_", " ").lower().capitalize(),
-                default=variables.metadata[key].default if key in variables.metadata else "",
-                type=str,
-            )
-        if user_input:
-            return user_input
-        else:
-            return (
-                user_input
-                if key in variables.metadata and variables.metadata[key].required
-                else None
-            )
+        if not (var := variables.get(key)):
+            raise ValueError(f"Unknown variable: {key!r}")
+
+        # Get the value from a source (cli/.../default), possibly prompting
+        # the user for a confirmation
+        with contextlib.suppress(KeyError):
+            return self._sources[key]
+
+        # No default value found, prompt user
+        click_prompt = partial(click.prompt, key.replace("_", " ").capitalize(), type=var.type)
+        if not var.required:
+            click_prompt.keywords["default"] = ""
+            return click_prompt() or None  # it asks only once
+        return click_prompt()  # it keeps asking until the user gives a value
+
+    def inject(self, key, value):
+        """Inject a confirmed key=value pair inside the environment."""
+        self._source_tool_confirmed[key] = value
```

Comment:
Now there's two way of injecting a value, we should only keep one:
```py
environ[key] = value
```
and
```py
environ.inject(key, value)
```

the former takes precedence on the cli/osenv, the latter does not. But this is not obvious and I'm not sure it will be more useful than confusing

Reply:
At least it should be documented to limit the confusion.